test/xbgpu/test_engine.py

```diff
@@ -236,116 +271,157 @@ def _create_heaps(
-        heaps. However, `batch_indices` is not required to contain full
-        accumulations.
+        Each full accumulation (for each corrprod-output) requires
+        `heap_accumulation_threshold` batches of heaps. However, `batch_indices`
+        is not required to contain full accumulations.
 
         Parameters
         ----------
         mock_recv_stream
             Fixture
-        recv_stream
-            InprocStream to receive data output by XBEngine.
+        mock_send_stream
+            Fixture
+        corrprod_outputs
+            Fixture
         heap_factory
             Callback that takes a batch index and returns the heaps for that index.
-        heap_accumulation_threshold
-            Number of consecutive heaps to process in a single accumulation.
         batch_indices
             Indices of the batches to send. These must be strictly increasing,
             but need not be contiguous.
         timestamp_step
             Timestamp step between each received heap processed.
-        n_ants, n_channels_per_substream, n_spectra_per_heap
+        n_ants, n_channels_per_substream, n_spectra_per_heap, frequency
             See :meth:`_create_heaps` for more info.
 
         Returns
         -------
         device_results
             Array of all GPU-generated data of shape
-            - (n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
+            - (len(corrprod_outputs), n_total_accumulations, n_channels_per_substream, n_baselines, COMPLEX)
+        n_accumulations_completed
+            Number of completed accumulations for each corrprod_output
         """
         max_packet_size = n_spectra_per_heap * N_POLS * COMPLEX * SAMPLE_BITWIDTH // 8 + PREAMBLE_SIZE
         max_heaps = n_ants * HEAPS_PER_FENGINE_PER_CHUNK * 10
-        feng_stream = self._make_feng(mock_recv_stream, max_packet_size, max_heaps)
+        feng_stream = self._make_feng(mock_recv_streams, max_packet_size, max_heaps)
+
+        acc_index_counts: list[dict[int, int]] = []
+        for corrprod_output in corrprod_outputs:
+            acc_index_counts.append(
+                dict.fromkeys(
+                    [batch_index // corrprod_output.heap_accumulation_threshold for batch_index in batch_indices], 0
+                )
+            )
+
+        # NOTE: The data generation generates an extra heap or several for the XPipeline
+        # with the smallest heap-accumulation-threshold - resulting in an incomplete final
+        # accumulation. Incomplete accumulations are purposefully dictated by `batch_indices`
+        # constructed by the caller.
+        for i, corrprod_output in enumerate(corrprod_outputs):
+            for batch_index in batch_indices:
+                adjusted_index = batch_index // corrprod_output.heap_accumulation_threshold
+                acc_index_counts[i][adjusted_index] += 1
+            # `adjusted_index` should be the last accumulation index for this corrprod
```

Comment:
That comment isn't helping me understand what the code below it is doing. It seems like if the last accumulation is incomplete you remove it from the count array? Why?